### PR TITLE
Allow the testing environment to perform request injection on ReqMgr2

### PR DESCRIPTION
**Pull request by ggonzr**
Related to: #1124

This enhances the testing environment to allow submissions to ReqMgr2 for integrating the test cases that involve injecting requests there. Some of the updates are summarized below:

1. Update the `action.sh` script to deploy an environment also for local development purposes, not relying only on the GitHub Action runners.
- In case some of the containers fail to start, this retrieves its log.
2. Retrieve all required credentials for opening SSH sessions from the environment. Also, mount certificates to authenticate requests to CMS Web services as secrets.
3. Allow the deployment of two versions, one for Python 2 and another for Python 3 to perform regression testing.
- Take the source from Dockerfile definitions directly from Git. Details [here](https://docs.docker.com/reference/compose-file/build/#attributes).
- Patch the CouchDB Lucene configuration file so that the Apache CouchDB source is not fixed.
- Provide the McM base URL dynamically, this is required to also integrate the `wmcontrol` module for injections.
4. Remove the dependency on AFS mounted in the deployment host. Use SSH sessions to prepare the working folder in AFS for submitting requests.
5. Remove the dependency on `sendmail`, and configure the SMTP session so that it can send emails on its own.
6. Remove duplicated code related to the initial test plan, and use directly the new version of the McM REST client by just installing the package.
7. Decrease the log level for some records to avoid flooding the log files.

Current working versions for Python 2 and Python 3 are available: [migration/current-python2](https://github.com/ggonzr/mcm/tree/migration/current-python2), [migration/current-python3](https://github.com/ggonzr/mcm/tree/migration/current-python3). After this PR is merged, I will rebase their content and push them into the upstream repository.

### CODE REVIEW

File: mcm/tools/ssh_executor.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import random
 
 from tools.logger import InjectionLogAdapter
+from tools.locator import locator
```
Comment (lmoureaux) [inline mcm/tools/ssh_executor.py:7]: Sort alphabetically:
```py
from tools.locator import locator
from tools.logger import InjectionLogAdapter
```

File: mcm/deploy/action.sh [hunk 2]
```diff
@@ -1,46 +1,101 @@
 #!/bin/bash
 
-# This is only intended to be executed by the
-# GitHub Action runner.
-
-# Sets up the environment and deploys the components
-echo "Docker version $(docker version)"
-echo "Docker Compose version $(docker compose version)"
+# Check Docker and Docker Compose versions
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:3]: This doesn't check the versions, only the availability.

File: mcm/deploy/action.sh [hunk 3]
```diff
@@ -1,46 +1,101 @@
...
+# Path to `couchdb-lucene.ini` required to set
+# this setting for the CouchDB Lucene component
+if [ -z "$TO_INI_PATH" ]; then
+    echo 'Set $TO_INI_PATH with the absolute path to the file `couchdb-lucene.ini` available in the repository for the execution environment'
+    exit 1
+fi
 
+# Set environment and context.
+DATE_WITH_TIME=`date "+%Y_%m_%d_%H_%M_%S"`
+TMP_FOLDER="/tmp/McM_Containers_Temporal_Data_Folder_${DATE_WITH_TIME}"
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:84]: Unsafe, use `mktemp`

File: mcm/deploy/action.sh [hunk 4]
```diff
@@ -1,46 +1,101 @@
...
+DATA_PATH="/tmp/McM_Container_Data_${DATE_WITH_TIME}"
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:85]: Unsafe, use `mktemp`

File: mcm/deploy/action.sh (hunk 3 above)
Comment (lmoureaux) [inline mcm/deploy/action.sh:84]: Temporal -> Temporary
(temporal = religious)

File: mcm/deploy/action.sh [hunk 5]
```diff
@@ -1,46 +1,101 @@
...
-echo "CouchDB Lucene"
-curl -s "http://localhost:$LUCENE_PORT/" | python3 -m json.tool
-echo "McM application"
-curl -s "http://localhost:8000/restapi/users/get_role" | python3 -m json.tool
+# Start the deployment
+function up() {
+    echo 'Starting deployment...'
+
+    # Create a temporal folder to download the sample data
+    rm -rf "${TMP_FOLDER}" && mkdir -p "${TMP_FOLDER}"
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:116]: Use `mktemp` (remove these lines; `mktemp` will create the folder)

File: mcm/deploy/action.sh [hunk 6]
```diff
@@ -1,46 +1,101 @@
...
+
+    # Download and decompress
+    echo 'Downloading McM data....'
+    curl -s $MCM_EXAMPLE_DATA_URL | tar -xzC "${TMP_FOLDER}/"
+
+    # Create a temporal folder to store the container's data
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:122]: temporal -> temporary
(temporal = religious)

File: mcm/deploy/action.sh [hunk 7]
```diff
@@ -1,46 +1,101 @@
...
+
+# Execute the procedure
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --up) up;;
+        --down) down;;
+        *) echo "Unknown option: $1"; exit 1;;
+    esac
+    shift
+done
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:185]: Newline at end of file (avoid mangling the prompt when using `cat`)

File: mcm/tests/mcm_tests/fixtures.py [hunk 8]
```diff
@@ -9,21 +9,30 @@
...
-from mcm_tests.rest_api.api_tools import Environment
+from mcm_tests.rest_api.api_tools import McMTesting, Environment, Roles
 from mcm_tests.use_cases.full_injection.core import InjectRootRequest, InjectToNanoAOD
 
 
 @pytest.fixture
-def stdin_enabled(request: FixtureRequest) -> None:
+def authenticate_by_stdin(request: FixtureRequest) -> bool:
+    """
+    Check if the user is going has enabled the `stdin` to authenticate
```
Comment (lmoureaux) [inline mcm/tests/mcm_tests/fixtures.py:19]: remove "is going"

File: mcm/deploy/couchdb-lucene.ini [hunk 9]
```diff
@@ -23,4 +23,4 @@ allowLeadingWildcard=false
 # couchdb server mappings
 
 [local]
-url = http://couchdb:5984/
\ No newline at end of file
+url=<MCM_COUCHDB_URL>
```
Comment (lmoureaux) [inline mcm/deploy/couchdb-lucene.ini:26]: Newline at end of file to avoid mangling the prompt when using `cat`

File: mcm/tools/installer.py [hunk 10]
```diff
@@ -54,10 +52,77 @@ def close(self):
...
+        """
+        Checks if a folder exists using the remote session.
+
+        Arguments:
+            path (str): Absolute path to check.
+
+        Returns:
+            bool: True if the folder exists, False otherwise.
+        """
+        _, stdout, stderr = self.ssh_session.execute("ls -l '%s'" % (path))
```
Comment (lmoureaux) [inline mcm/tools/installer.py:69]: Open an SFTP session and use [listdir()](https://docs.paramiko.org/en/latest/api/sftp.html#paramiko.sftp_client.SFTPClient.listdir):
```py
self.ssh_client.open_sftp().listdir(path)
```

File: mcm/tools/installer.py [hunk 11]
```diff
@@ -54,10 +52,77 @@ def close(self):
...
+        if output and not error:
+            # Most probably we found a valid path
+            return True
+        if error and not output:
+            # Check the proper message is in the `stderr`
+            if "No such file or directory" in error:
+                return False
+            raise RuntimeError("Unexpected message for checking if a folder exists: %s" % (error))
+        
+    def _create_folder(self, path):
```
Comment (lmoureaux) [inline mcm/tools/installer.py:86]: [mkdir()](https://docs.paramiko.org/en/latest/api/sftp.html#paramiko.sftp_client.SFTPClient.mkdir) on the SFTP session

File: mcm/tools/installer.py [hunk 12]
```diff
@@ -54,10 +52,77 @@ def close(self):
...
+            raise RuntimeError(msg)
+        
+        error = stderr.read()
+        if error:
+            raise RuntimeError("Unable to create the folder: %s - Description: %s" % (path, error))
+
+        if not self._folder_exists(path):
+            raise RuntimeError("Folder not found (%s), it should exist" % (path))
+        
+    def _remove_folder(self, path):
```
Comment (lmoureaux) [inline mcm/tools/installer.py:106]: [rmdir()](https://docs.paramiko.org/en/latest/api/sftp.html#paramiko.sftp_client.SFTPClient.rmdir) on the SFTP session

File: mcm/tools/communicator.py [hunk 13]
```diff
@@ -17,6 +17,11 @@ class communicator:
     def __init__(self):
         self.from_opt = 'user'  # could be service at some point
 
+    def _smtp_session(self):
+        smtp = smtplib.SMTP(host="cernmx.cern.ch", port=25)
```
Comment (lmoureaux) [inline mcm/tools/communicator.py:21]: Read from the environment

File: mcm/tools/communicator.py [hunk 14]
```diff
@@ -41,8 +46,7 @@ def flush(self, Nmin):
                 # self.logger.info('Sending a message from cache \n%s'% (text))
                 try:
                     msg.attach(MIMEText(text))
-                    smtpObj = smtplib.SMTP()
-                    smtpObj.connect()
+                    smtpObj = self._smtp_session()
                     smtpObj.sendmail(sender, destination, msg.as_string())
                     smtpObj.quit()
```
Comment (lmoureaux) [inline mcm/tools/communicator.py:51]: While you're at it:
```py
with self._smtp_session() as smtp:
    smtp.sendmail(sender, destination, msg.as_string())
```

File: mcm/tools/communicator.py [hunk 15]
```diff
@@ -114,8 +118,7 @@ def sendMail(self,
 
         try:
             msg.attach(MIMEText(text))
-            smtpObj = smtplib.SMTP()
-            smtpObj.connect()
+            smtpObj = self._smtp_session()
             communicator.logger.info('Sending %s to %s...' % (msg['Subject'], msg['To']))
             smtpObj.sendmail(sender, destination, msg.as_string())
             smtpObj.quit()
```
Comment (lmoureaux) [inline mcm/tools/communicator.py:124]: Context manager:
```py
with self._smtp_session() as smtp:
    communicator.logger.info('Sending %s to %s...' % (msg['Subject'], msg['To']))
    smtp.sendmail(sender, destination, msg.as_string())
```

**Review by lmoureaux — COMMENTED** [review]
Functionally looks good, I have some style comments

File: mcm/deploy/action.sh (hunk 2 above)
Comment (ggonzr) [reply mcm/deploy/action.sh:3]: Sure, the error message below raises a message about this. However, if the execution finishes successfully, it allows us to know which version is available for both components. Could you clarify the purpose of this comment?

File: mcm/tools/installer.py (hunk 12 above)
Comment (ggonzr) [reply mcm/tools/installer.py:106]: The folder will have some files inside, so just calling `rmdir()` will raise an exception here.

File: mcm/tools/installer.py (hunk 11 above)
Comment (ggonzr) [reply mcm/tools/installer.py:86]: Likewise for this, it is not possible to emulate the `mkdir -p` instruction just with a call of `mkdir()` on the SFTP session. So I suggest to leave this as it is.

File: mcm/deploy/action.sh (hunk 2 above)
Comment (lmoureaux) [reply mcm/deploy/action.sh:3]: Just that the comment is misleading

File: mcm/tools/installer.py (hunk 12 above)
Comment (lmoureaux) [reply mcm/tools/installer.py:106]: ok fair enough

File: mcm/tools/installer.py (hunk 11 above)
Comment (lmoureaux) [reply mcm/tools/installer.py:86]: Fair enough

File: mcm/tools/installer.py [hunk 16]
```diff
@@ -66,22 +66,14 @@ def _folder_exists(self, path):
...
             return True
-        if error and not output:
-            # Check the proper message is in the `stderr`
-            if "No such file or directory" in error:
+        except IOError as e:
+            if e.errno == errno.ENOENT:
+                # Folder doesn't exists
                 return False
-            raise RuntimeError("Unexpected message for checking if a folder exists: %s" % (error))
+            raise e
```
Comment (lmoureaux) [inline mcm/tools/installer.py:76]: Could be just `raise` without specifying the exception

**Review by lmoureaux — APPROVED** [review]

**Pull request merged**